clippy_lints/src/module_style.rs

```diff
@@ -0,0 +1,163 @@
+use std::{
+    fmt,
+    path::{Component, Path},
+};
+
+use rustc_ast::ast;
+use rustc_data_structures::fx::FxHashSet;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{FileName, RealFileName};
+use serde::{de, Deserialize};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks that module layout is consistent with configured style.
+    ///
+    /// ### Why is this bad?
+    /// Having multiple module layout styles in a project can be confusing.
+    ///
+    /// ### Example
+    /// ```text
+    /// src/
+    ///   stuff/
+    ///     stuff_files.rs
+    ///     mod.rs
+    ///   lib.rs
+    /// ```
+    /// Use instead:
+    /// ```text
+    /// src/
+    ///   stuff/
+    ///     stuff_files.rs
+    ///   stuff.rs
+    ///   lib.rs
+    /// ```
+    pub MODULE_STYLE,
+    style,
+    "checks that module layout is consistent"
+}
+
+pub struct ModStyle {
+    config: ModConfig,
+}
+
+impl ModStyle {
+    pub fn new(config: ModConfig) -> Self {
+        Self { config }
+    }
+}
+
+impl_lint_pass!(ModStyle => [MODULE_STYLE]);
+
+impl EarlyLintPass for ModStyle {
+    fn check_crate(&mut self, cx: &EarlyContext<'_>, _: &ast::Crate) {
+        let files = cx.sess.source_map().files();
+        match self.config {
+            ModConfig::ModuleFile => {
+                let trim_to_src = if let RealFileName::LocalPath(p) = &cx.sess.working_dir {
+                    p.to_string_lossy()
+                } else {
+                    return;
+                };
+
+                let mut paths = FxHashSet::default();
+                let mut num_mod_files = 0;
+                for file in files.iter().map(|sf| &sf.name) {
+                    match file {
+                        FileName::Real(RealFileName::LocalPath(p))
+                            // TODO                      this is hacky
+                            if p.to_string_lossy().starts_with(trim_to_src.as_ref()) =>
+                        {
+                            let p = p.to_string_lossy();
+                            let path = Path::new(p.trim_start_matches(trim_to_src.as_ref()));
+
+                            if path.ends_with("mod.rs") {
```

Comment:
This should not trigger for `something_mod.rs`.

I would just emit the lint right here. It's okay to emit for every module. Also, you should be able to get a `Span` using `Span::new(file.start_pos, file.start_pos, SyntaxContext::root())`.

Reply:
It shouldn't `Path::ends_with` is different that `str::ends_with`, it only considers whole segments. 

I can't emit until I know if there are any missing `mod.rs` files this is the `mod_file` check. I could still collect all the folder names (after `src`) and collect all the mod.rs file paths and remove all the folders that contain mod files, emitting for all the folders that are left after. I would have to keep more info around and haven't figured out all the details but it would probably work, then the error messages would at least be more consistent.